python/tchannel/tornado/hyperbahn.py

```diff
@@ -58,34 +120,33 @@ def advertise(tchannel, service, routers):
         # TChannel already knows about some of the routers.
         tchannel.peers.get(router)
 
-    @tornado.gen.coroutine
-    def _register():
-        response = yield tchannel.request(service='hyperbahn').send(
-            arg1='ad',  # advertise
-            arg2='',
-            arg3=json.dumps({
-                'services': [
-                    {
-                        'serviceName': service,
-                        'cost': 0,
-                    }
-                ]
-            }),
-            headers={'as': 'json'},
-        )
-
-        if response.code not in ErrorCode:
-            # re-register every ``DEFAULT_TTL`` seconds
-            tornado.ioloop.IOLoop.current().call_later(
-                delay=DEFAULT_TTL,
-                callback=_register,
+    # first advertise rule apply here.
+    ad_future = Future()
```

Comment:
Ok so we could just raise the exception explicitly:

```
while True:
    if been_more_than_30s:
        raise SomeError()
```

The big idea here is that right now `_advertise` is trying to *manage itself*, and we need to move that management into `advertise`.